sdk/dslink/src/main/java/org/dsa/iot/dslink/link/SubscriptionHelper.java

```diff
@@ -0,0 +1,192 @@
+ * path are not possible.  This uses a multiplexing adapter so that the requester
+ * thinks there is only one subscriber.
+ *
+ * @author Aaron Hansen
+ */
+public class SubscriptionHelper {
+
+    // Fields
+    // ------
+
+    private Requester requester;
+    private ConcurrentHashMap<String, Adapter> subscriptions;
+
+    // Constructors
+    // ------------
+
+    public SubscriptionHelper(Requester requester) {
+        if (requester == null) {
+            throw new NullPointerException("Requester must not be null");
+        }
+        this.requester = requester;
+        subscriptions = new ConcurrentHashMap<String, Adapter>();
+    }
+
+
+    // Public Methods
+    // --------------
+
+    /**
+     * Clears all subscriptions, but does not call unsubscribe on the requester.
+     */
+    public void clear() {
+        if (subscriptions != null) {
+            subscriptions.clear();
+        }
+    }
+
+    /**
+     * Clears a single subscription without calling unsubscribe on the requester.
+     */
+    public SubscriptionHelper clear(String path) {
+        Adapter a = subscriptions.remove(path);
+        if (a != null) {
+            requester.unsubscribe(path, null);
+        }
+        return this;
+    }
+
+    /**
+     * Safely subscribes the path, even if it is already subscribed.  Calls
+     * subscribe with a SubData for the given path and a QOS of 0.
+     */
+    public SubscriptionHelper subscribe(String path,
+                                        Handler<SubscriptionValue> handler) {
+        return subscribe(new SubData(path, 0), handler);
+    }
+
+    /**
+     * Safely subscribes the path, even if it is already subscribed.  The path will
+     * be subscribed at the give QOS of the very first subscription.  Subsequent
+     * subscriptions will get the same QOS.
+     *
+     * @param path The first scription
+     */
+    @SuppressFBWarnings("AT_OPERATION_SEQUENCE_ON_CONCURRENT_ABSTRACTION")
+    public SubscriptionHelper subscribe(SubData path,
+                                        Handler<SubscriptionValue> handler) {
+        Adapter a = subscriptions.get(path.getPath());
+        if (a == null) {
+            Adapter adapter = new Adapter();
+            adapter.add(handler);
+            requester.subscribe(path, adapter);
+            subscriptions.put(path.getPath(), adapter);
+        } else {
+            a.add(handler);
+        }
+        return this;
+    }
+
+    /**
+     * Only unsubscribes the given handler.
+     */
+    @SuppressFBWarnings("AT_OPERATION_SEQUENCE_ON_CONCURRENT_ABSTRACTION")
+    public SubscriptionHelper unsubscribe(String path,
+                                          Handler<SubscriptionValue> handler) {
+        Adapter a = subscriptions.get(path);
+        if (a != null) {
+            a.remove(handler);
+            if (a.size() == 0) {
+                requester.unsubscribe(path, null);
```

Comment:
If I'm not mistaken, you could pass a closure here, to at least log it if an error happens during the unsubscription. It could help during debugging.